src/main/java/org/wildfly/naming/client/remote/RemoteClientTransport.java

```diff
@@ -205,13 +206,28 @@ Object lookup(final RemoteContext context, final Name name, final boolean preser
             }
             final BlockingInvocation.Response response = invocation.getResponse();
             try (MessageInputStream is = response.getInputStream()) {
-                if (is.readUnsignedByte() == Protocol.P_CONTEXT) {
-                    return new RelativeFederatingContext(new FastHashtable<>(context.getEnvironment()), context, NamingUtils.toCompositeName(name));
-                } else if (is.readUnsignedByte() != Protocol.P_OBJECT) {
-                    throw Messages.log.invalidResponse();
-                }
-                try (Unmarshaller unmarshaller = createUnmarshaller(is)) {
-                    return unmarshaller.readObject();
+                final int result = is.readUnsignedByte();
+                if (result == Protocol.SUCCESS) {
```

Comment:
In this case each return item is expected to be one of the valid P_ (parameter) types.  P_EXCEPTION implies failure, whereas P_CONTEXT and P_OBJECT imply success.  I think the real bug is that the original response handler wasn't ready to handle exceptions.